server/testdriver/core/src/main/java/org/infinispan/server/test/core/rollingupgrade/RollingUpgradeHandler.java

```diff
@@ -0,0 +1,229 @@
+
+   private RemoteCacheManager remoteCacheManager;
+
+   private STATE currentState = STATE.NOT_STARTED;
+
+   public enum STATE {
+      /**
+       * The upgrade procedure has not been started yet
+       */
+      NOT_STARTED,
+      /**
+       * This state signals that all the old version nodes are running currently
+       */
+      OLD_RUNNING,
+      /**
+       * This state signals that one old node has been shut down. The remaining nodes may be a mixture of old and/or new
+       */
+      REMOVED_OLD,
+      /**
+       * This state signals that a new node was just added. The remaining nodes may be a mixture of old and/or new
+       */
+      ADDED_NEW,
+      /**
+       * This state signals that the upgrade completed and all old nodes have been removed and all new nodes are running
+       */
+      NEW_RUNNING,
+      /**
+       * The upgrade procedure encountered some type of error
+       */
+      ERROR
+   }
+
+   private RollingUpgradeHandler(RollingUpgradeConfiguration configuration) {
+      this.configuration = configuration;
+      this.logConsumer = configuration.logConsumer();
+   }
+
+   public String getToImageCreated() {
+      return toImageCreated;
+   }
+
+   public String getFromImageCreated() {
+      return fromImageCreated;
+   }
+
+   public ContainerInfinispanServerDriver getFromDriver() {
+      return fromDriver;
+   }
+
+   public ContainerInfinispanServerDriver getToDriver() {
+      return toDriver;
+   }
+
+   public RollingUpgradeConfiguration getConfiguration() {
+      return configuration;
+   }
+
+   public RemoteCacheManager getRemoteCacheManager() {
+      return remoteCacheManager;
+   }
+
+   public STATE getCurrentState() {
+      return currentState;
+   }
+
+   public static void performUpgrade(RollingUpgradeConfiguration configuration) throws InterruptedException {
+      String site1Name = "site1";
+      RollingUpgradeHandler handler = new RollingUpgradeHandler(configuration);
+
+      int nodeCount = configuration.nodeCount();
+      String versionFrom = configuration.fromVersion();
+      String versionTo = configuration.toVersion();
+      try {
+         handler.logConsumer.accept("Starting " + nodeCount + " node to version " + versionFrom);
+         handler.fromDriver = handler.startNode(false, configuration.nodeCount(), configuration.nodeCount(),
+               site1Name, configuration.jgroupsProtocol(), null);
+         handler.currentState = STATE.NEW_RUNNING;
+
+         try (RemoteCacheManager manager = handler.createRemoteCacheManager()) {
+            handler.remoteCacheManager = manager;
+            configuration.initialHandler().accept(handler);
+
+            for (int i = 0; i < nodeCount; ++i) {
+               handler.logConsumer.accept("Shutting down 1 node from version: " + versionFrom);
+               int nodeId = nodeCount - i - 1;
+               String volumeId = handler.fromDriver.volumeId(nodeId);
+               handler.fromDriver.stop(nodeId);
```

Comment:
It might be a silly question: what would be the difference between starting the new and stopping the old one? It wouldn't be possible to start from the same volume. Is there something else? Then we should also emphasize always stopping the previous and then starting the new version.